examples/esp32-with-cred/esp32-server-with-cred.rs

```diff
@@ -0,0 +1,246 @@
+    micro_rdk::common::registry::ComponentRegistry,
+};
+
+#[derive(Debug, Error)]
+pub enum ServerError {
+    #[error("Error fetching NVS key: {0}")]
+    NVSKeyError(String),
+    #[error("{0}")]
+    EspError(EspError),
+    #[error("Error obtaining peripherals")]
+    PeripheralsError,
+}
+
+impl From<EspError> for ServerError {
+    fn from(value: EspError) -> ServerError {
+        ServerError::EspError(value)
+    }
+}
+
+const VIAM_NVS_NAMESPACE: &str = "VIAM_NS";
+
+#[cfg(not(feature = "qemu"))]
+fn get_str_from_nvs(viam_nvs: &EspDefaultNvs, key: &str) -> Result<String, ServerError> {
+    let mut buffer_ref = [0_u8; 4000];
+    Ok(viam_nvs
+        .get_str(key, &mut buffer_ref)?
+        .ok_or(ServerError::NVSKeyError(key.to_string()))?
+        .trim_matches(char::from(0))
+        .to_string())
+}
+
+#[cfg(not(feature = "qemu"))]
+fn get_blob_from_nvs(viam_nvs: &EspDefaultNvs, key: &str) -> Result<Vec<u8>, ServerError> {
+    let mut buffer_ref = [0_u8; 4000];
+    Ok(viam_nvs
+        .get_blob(key, &mut buffer_ref)?
+        .ok_or(ServerError::NVSKeyError(key.to_string()))?
+        .to_vec())
+}
+
+#[cfg(not(feature = "qemu"))]
+struct NvsStaticVars {
+    wifi_ssid: String,
+    wifi_pwd: String,
+    robot_secret: String,
+    robot_id: String,
+    robot_dtls_cert: Vec<u8>,
+    robot_dtls_key_pair: Vec<u8>,
+    robot_dtls_cert_fp: String,
+    robot_srv_pem_chain: Vec<u8>,
+    robot_srv_pem_ca: Vec<u8>,
+    robot_srv_der_key: Vec<u8>,
+}
+
+impl NvsStaticVars {
+    fn new() -> Result<NvsStaticVars, ServerError> {
+        let nvs = EspDefaultNvsPartition::take()?;
+        info!("get namespace...");
+        let viam_nvs = EspNvs::new(nvs.clone(), VIAM_NVS_NAMESPACE, true)?;
+        info!("loading creds...");
+        let res = Ok(NvsStaticVars {
+            wifi_ssid: get_str_from_nvs(&viam_nvs, "WIFI_SSID")?,
+            wifi_pwd: get_str_from_nvs(&viam_nvs, "WIFI_PASSWORD")?,
+            robot_secret: get_str_from_nvs(&viam_nvs, "ROBOT_SECRET")?,
+            robot_id: get_str_from_nvs(&viam_nvs, "ROBOT_ID")?,
+            robot_dtls_cert: get_blob_from_nvs(&viam_nvs, "ROBOT_DTLS_CERT")?,
+            robot_dtls_key_pair: get_blob_from_nvs(&viam_nvs, "DTLS_KEY_PAIR")?,
+            robot_dtls_cert_fp: get_str_from_nvs(&viam_nvs, "DTLS_CERT_FP")?,
+            robot_srv_pem_chain: get_blob_from_nvs(&viam_nvs, "SRV_PEM_CHAIN")?,
+            robot_srv_pem_ca: get_blob_from_nvs(&viam_nvs, "CA_CRT")?,
+            robot_srv_der_key: get_blob_from_nvs(&viam_nvs, "SRV_DER_KEY")?,
+        });
+        res
+    }
+}
+
+fn main() -> Result<(), ServerError> {
+    std::env::set_var("RUST_BACKTRACE", "1");
+    esp_idf_sys::link_patches();
+
+    esp_idf_svc::log::EspLogger::initialize_default();
+    let sys_loop_stack = EspSystemEventLoop::take()?;
+
+    #[cfg(not(feature = "qemu"))]
+    let periph = Peripherals::take().ok_or(ServerError::PeripheralsError)?;
+
+    #[cfg(feature = "qemu")]
```

Comment:
There are enough differences in here that I think you should test this out both with and without qemu, to make sure it works both ways.

Reply:
I agree, there are some bugs I'm noticing here now that I would have caught if I had